packages/plugin-ext/src/main/browser/scm-main.ts

```diff
@@ -37,12 +37,29 @@ export class ScmMainImpl implements ScmMain {
     private readonly scmService: ScmService;
     private readonly scmRepositoryMap: Map<number, ScmRepository>;
     private readonly labelProvider: LabelProvider;
+    private selectedSourceControlHandle: number | undefined;
 
     constructor(rpc: RPCProtocol, container: interfaces.Container) {
         this.proxy = rpc.getProxy(MAIN_RPC_CONTEXT.SCM_EXT);
         this.scmService = container.get(ScmService);
         this.scmRepositoryMap = new Map();
         this.labelProvider = container.get(LabelProvider);
+        this.scmService.onDidChangeSelectedRepository(repository => {
+            let externalRepository: boolean = false;
+            if (repository) {
+                this.scmRepositoryMap.forEach((value, key) => {
+                    if (value.provider.rootUri === repository.provider.rootUri) {
+                        externalRepository = true;
+                        this.selectedSourceControlHandle = key;
+                        this.proxy.$setSourceControlSelection(key, true);
+                    }
+                });
+            }
+            if (!externalRepository && this.selectedSourceControlHandle !== undefined) {
```

Comment:
Confused about `externalRepository` - does it mean something selected? If so should not we unselect previous selection anyway, i.e.
```
const sourceControlHandle = this.getSourceControlHandle(repository);
if (sourceControlHandle !== this.selectedSourceControlHandle) {
   this.proxy.$setSourceControlSelection(this.selectedSourceControlHandle, sourceControlHandle); // old, new, both of `number | undefined` type
   this.selectedSourceControlHandle = sourceControlHandle;
}

Reply:
`externalRepository` means that the selected repository is registered from the plugin, so it has a handle id (the repository exists in the `scmRepositoryMap`). The repository may not be present in the `scmRepositoryMap` if it is registered from internal extension e.g. built in Git extension. In this  case we need to deselect previously selected repository that has a handle id.

Reply:
I cannot understand whether `this.proxy.$setSourceControlSelection(this.selectedSourceControlHandle, false);` will be called in the case if `this.proxy.$setSourceControlSelection(key, true);` is called above. From code it looks like it won't because `!externalRepository`. So it means that the previous source control won't be unselected.

Reply:
I also don't get why internal/external is important. I thought the logic should be unselect last handle if it is not the same as current. If repository is internal then handle will be undefined and we select nothing, but still unselect the previous handle.

Reply:
Reworked the logic

Reply:
the logic is straight to understand now, thank you!